src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -93,47 +121,86 @@ void Bank::evaluate_and_allocate() {
+                const auto& lt = vec[idx];
+                std::unique_lock dev_guard(device_bank.mutex);
+                auto iter_device_registered = device_bank.registered_tensors.find(lt);
+                NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                            "Tensor should be registered first!");
+                auto uid = iter_device_registered->second;
+                if (device_bank.storage[uid].tensor) {
+                    // Already allocated
+                    return;
+                }
+
+                dev_guard.unlock();
+                auto transformed = lt.eval();
+
+                // Note: No need to relock here as the storage is already prepared for writing
+                device_bank.storage.at(uid).tensor = std::move(transformed);
+                const_cast<LazyTensor&>(lt).detach();
+            });
+        } else {
+            // Allocation needed
             std::unique_lock dev_guard(device_bank.mutex);
-            auto iter_device_registered = device_bank.registered_tensors.find(lt);
-            NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
-                        "Tensor should be registered first!");
-            if (device_bank.storage[iter_device_registered->second].tensor) {
-                // Already allocated
-                return;
+            for (std::size_t idx = 0; idx < vec.size(); ++idx) {
+                const auto& lt = vec[idx];
+                auto iter_device_registered = device_bank.registered_tensors.find(lt);
+                NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
+                            "Tensor should be registered first!");
+                auto uid = iter_device_registered->second;
+                if (device_bank.storage[uid].tensor) {
+                    // Already allocated
+                    continue;
+                }
+
+                uids_to_allocated[uid] = {true, lt.eval_meta(), ov::Tensor()};
             }
-            dev_guard.unlock();
-
-            // Allocation and/or evaluation needed
-            // Evaluate concurrently, lock the device
-            // mutex only to update the device bank (& allocate on-device memory, if needed)
-            const auto& transformed_tensor = lt.eval();
+        }
 
-            dev_guard.lock();
-            if (device_for_alloc == "CPU") {
-                // No allocation needed
-                device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = transformed_tensor;
-                const_cast<LazyTensor&>(lt).detach();
-                return;
+        std::unique_lock guard(device_bank.mutex);
+        // Allocate memory sequentially - in order of UID
+        for (std::size_t i = 0; i < uids_to_allocated.size(); ++i) {
+            auto& allocated = uids_to_allocated[i];
+            if (!allocated.needs_alloc) {
+                continue;
```

Comment:
you never seem to push `false` to that tensor, is that right?

Reply:
Right, reworked